### ISSUE TITLE
[BUG] gpu_package seems not to follow OpenCL 2.0 and does not compile with ROCm 2.2

### ISSUE THREAD

**Issue by vvsteg** [lammps/lammps#1368]
The current version of LAMMPS is compiled with the opencl variant of the gpu_package under Ubuntu 18.04 (14.15 kernel) with ROCm 2.2. There is one Radeon VII GPU in the system.

The standard ```example/melt``` benchmark provides the following errors messages during execution
```
../../src/lmp_mpi -sf gpu -pk gpu 1 -in in.melt 
LAMMPS (28 Feb 2019)
Lattice spacing in x,y,z = 1.6796 1.6796 1.6796
Created orthogonal box = (0 0 0) to (67.1838 67.1838 67.1838)
  1 by 1 by 1 MPI processor grid
Created 256000 atoms
  Time spent = 0.0260232 secs

--------------------------------------------------------------------------
- Using acceleration for lj/cut:
-  with 1 proc(s) per device.
-  with OpenCL Parameters for: GENERIC
--------------------------------------------------------------------------
Device 0: gfx906, 60 CUs, 16 GB, 1.8 GHZ (Mixed Precision)
--------------------------------------------------------------------------

Initializing Device and compiling on process 0.../tmp/AMD_3903_69/t_3903_71.cl:582:17: error: variables in the local address space can only be declared in the outermost scope of a kernel function
    __local int n_stride;
                ^
/tmp/AMD_3903_69/t_3903_71.cl:621:5: error: variables in the local address space can only be declared in the outermost scope of a kernel function
    store_answers(f,energy,virial,ii,inum,tid,t_per_atom,offset,eflag,vflag,
    ^
/tmp/AMD_3903_69/t_3903_71.cl:395:20: note: expanded from macro 'store_answers'
    __local acctyp red_acc[6][BLOCK_PAIR];                                  \
                   ^
/tmp/AMD_3903_69/t_3903_71.cl:656:17: error: variables in the local address space can only be declared in the outermost scope of a kernel function
    __local int n_stride;
                ^
/tmp/AMD_3903_69/t_3903_71.cl:694:5: error: variables in the local address space can only be declared in the outermost scope of a kernel function
    store_answers(f,energy,virial,ii,inum,tid,t_per_atom,offset,eflag,vflag,
    ^
/tmp/AMD_3903_69/t_3903_71.cl:395:20: note: expanded from macro 'store_answers'
    __local acctyp red_acc[6][BLOCK_PAIR];                                  \
                   ^
4 errors generated.
Done.
Initializing Device 0 on core 0...Done.

Setting up Verlet run ...
  Unit style    : lj
  Current step  : 0
  Time step     : 0.005
OpenCL error in file 'geryon/ocl_kernel.h' in line 219 : -48.
```
If I am not mistaken the reason of these error messages is that the gpu_package does not follow the OpenCL C specifications required by OpenCL 2.0. 

See a similar problem discussed here [https://github.com/RadeonOpenCompute/ROCm/issues/674](url)

**Comment by akohlmey** [comment]
@vvsteg the OpenCL kernels in LAMMPS predate OpenCL 2.0 and have been in limited maintenance mode since. Can you try if forcing OpenCL 1.2 compatibility at run time with the change below would work around the reported issue? As an alternative, you may look into KOKKOS, which seems to support ROCm now and is more actively maintained.

```
  diff --git a/lib/gpu/lal_device.cpp b/lib/gpu/lal_device.cpp
  index 6b4d0ab2a..9397f3c6c 100644
  --- a/lib/gpu/lal_device.cpp
  +++ b/lib/gpu/lal_device.cpp
  @@ -237,7 +237,7 @@ int DeviceT::set_ocl_params(char *ocl_vendor) {
                           " -DBLOCK_CELL_ID="+params[11]+
                           " -DMAX_BIO_SHARED_TYPES="+params[12];
     }
  -  _ocl_compile_string="-cl-fast-relaxed-math -cl-mad-enable "+std::string(OCL_INT_TYPE)+" "+
  +  _ocl_compile_string="-cl-std=CL1.2 -cl-fast-relaxed-math -cl-mad-enable "+std::string(OCL_INT_TYPE)+" "+
                         std::string(OCL_PRECISION_COMPILE)+" "+_ocl_vendor_string;
     #endif
     return 0;
```

**Comment by vvsteg** [comment]
Thank you very much for the hints!

I have tried the variant with ```-cl-std=CL1.2```. No success. The error messages have not changed. I am going to post about this compatibility issue at the ROCm list.

I have made an attempt to build Kokkos for ROCm (without MPI) using these setting
```
CC =            /opt/rocm/hcc/bin/clang++
CCFLAGS =       -g -O3
SHFLAGS =       -fPIC
DEPFLAGS =      -M

LINK =          /opt/rocm/hcc/bin/clang++
LINKFLAGS =     -g -O3
LIB =
SIZE =          size

ARCHIVE =       ar
ARFLAGS =       -rc
SHLIBFLAGS =    -shared
KOKKOS_DEVICES = ROCm
```
Strangely enough, it seems that C++ AMP library is used in KOKKOS for ROCm. The compilation fails for ROCm v2.2 since the AMP support has been depreciated in ROCm after version 1.9:
```
/opt/rocm/hcc/bin/clang++ -g -O3  -DLAMMPS_GZIP -DLMP_KOKKOS  -I../STUBS    -I./ -I../../lib/kokkos/core/src -I../../lib/kokkos/containers/src -I../../lib/kokkos/algorithms/src -I../../lib/kokkos/core/src/eti   -hc -std=c++amp -I/opt/rocm/hcc/include -I/opt/rocm/include  --gcc-toolchain=/usr   -c ../fix_external.cpp
/opt/rocm/hcc/bin/clang++ -g -O3  -DLAMMPS_GZIP -DLMP_KOKKOS  -I../STUBS    -I./ -I../../lib/kokkos/core/src -I../../lib/kokkos/containers/src -I../../lib/kokkos/algorithms/src -I../../lib/kokkos/core/src/eti   -hc -std=c++amp -I/opt/rocm/hcc/include -I/opt/rocm/include  --gcc-toolchain=/usr   -c ../math_special.cpp
/opt/rocm/hcc/bin/clang++ -g -O3  -DLAMMPS_GZIP -DLMP_KOKKOS  -I../STUBS    -I./ -I../../lib/kokkos/core/src -I../../lib/kokkos/containers/src -I../../lib/kokkos/algorithms/src -I../../lib/kokkos/core/src/eti   -hc -std=c++amp -I/opt/rocm/hcc/include -I/opt/rocm/include  --gcc-toolchain=/usr   -c ../finish.cpp
In file included from ../finish.cpp:22:
In file included from ../accelerator_kokkos.h:22:
In file included from ../kokkos.h:18:
In file included from ../kokkos_type.h:19:
In file included from ../../lib/kokkos/core/src/Kokkos_Core.hpp:78:
../../lib/kokkos/core/src/Kokkos_ROCm.hpp:75:10: fatal error: 'amp_math.h' file not found
#include <amp_math.h>
         ^~~~~~~~~~~~
1 error generated.
In file included from ../finish.cpp:22:
In file included from ../accelerator_kokkos.h:22:
In file included from ../kokkos.h:18:
In file included from ../kokkos_type.h:19:
In file included from ../../lib/kokkos/core/src/Kokkos_Core.hpp:78:
../../lib/kokkos/core/src/Kokkos_ROCm.hpp:75:10: fatal error: 'amp_math.h' file not found
#include <amp_math.h>
         ^~~~~~~~~~~~
1 error generated.
Makefile:109: recipe for target 'finish.o' failed
make[1]: *** [finish.o] Error 1
make[1]: Leaving directory '/home/jiht-202-over-pc/lammps/src/Obj_kokkos_rocm'
Makefile:202: recipe for target 'kokkos_rocm' failed
make: *** [kokkos_rocm] Error 2
```

**Comment by stanmoore1** [comment]
@crtrott or @chipfreitag can you comment on building Kokkos with ROCm? I've never done this since I don't have an AMD GPU.

**Comment by akohlmey** [comment]
> Thank you very much for the hints!

that is the best i could do from what i found through searching with google. :-(
 
> I have tried the variant with `-cl-std=CL1.2`. No success. The error messages have not changed. I am going to post about this compatibility issue at the ROCm list.

perhaps @ndtrung81 or @wmbrownIntel have some suggestion or can point out a mistake in my hack.

**Comment by chipfreitag** [comment]
@rwvo resolved the amp_math.h include problem recently and will be working with the Kokkos team to resolve that issue.  There are a couple of others we need to work through also. This will get  Kokkos ROCm to the point where it once again builds and run tests.

However.  We know that that lammps will not build with the current ROCm backend.  We have a plan to resolve this and will be putting some resources on it.
The OpenCL path is more likely to yield a working program in the short term.

**Comment by ndtrung81** [comment]
The compiling errors in the original post indicated that the local space variables should be in the outermost level of the kernel. Can you try move red_acc out of the check in lal_aux_fun1.h, and also remove __local for n_stride in lal_lj.cu (as below), and rebuild the GPU library with OpenCL 2.0 and then LAMMPS, and rerun the test with in.melt?

diff lal_aux_fun1.h 
```
47d46
<   __local acctyp red_acc[6][BLOCK_PAIR];                                    \
48a48
>     __local acctyp red_acc[6][BLOCK_PAIR];                                  \
93d92
<   __local acctyp red_acc[6][BLOCK_PAIR];                                    \
94a94
>     __local acctyp red_acc[6][BLOCK_PAIR];                                  \
```
diff lal_lj.cu
```
50c50
<     int n_stride;
---
>     __local int n_stride;
139c139
<     int n_stride;
---
>     __local int n_stride;

```
If these changes indeed resolve the issue, similar changes can be made to other pair styles (lal_*.cu files).

**Comment by vvsteg** [comment]
@akohlmey, @chipfreitag, @ndtrung81 thank you for the information and the hints!

I have implemented the modification of the ```__local``` variable definitions proposed by @ndtrung81. Unfortunately, these small changes are not enough. 

If one just moves ``` __local acctyp red_acc[6][BLOCK_PAIR];``` before the if-statement of the ```store_answers``` macro in ```lal_aux_fun1.h```, OpenCL runtime still complains that ```variables in the local address space can only be declared in the outermost scope of a kernel function```.

I have been successful in getting a working LAMMPS for a generic LJ model after I copied this macro directly into the ```k_lj``` and ```k_lj_fast``` kernels in ```lal_lj.cu```.

I am not sure if this copy-paste method for the ```store_answers``` macro is the best way to get OpenCL 2.0 compliant code.

Btw, are there any evident expected difficulties on the way to get a hipified CUDA-version of the GPU-pakage? Perhaps, it could be a more efficient method for porting LAMMPS to ROCm infrastructure?

**Comment by rbberger** [comment]
I can confirm this is also an issue with Intel's Compute Runtime, which is Intel's new OpenCL driver stack, which is also OpenCL 2.0.

**Comment by akohlmey** [comment]
> Btw, are there any evident expected difficulties on the way to get a hipified CUDA-version of the GPU-pakage? Perhaps, it could be a more efficient method for porting LAMMPS to ROCm infrastructure?

as it so happens, i was contacted about a week ago by somebody, who claims to have ported the geryon library to ROCm HIP. this will hopefully be merged in the near future. Also it would result in ~20% speedup vs. OpenCL

**Comment by ndtrung81** [comment]
Glad to hear that!

**Comment by vvsteg** [comment]
@akohlmey, actually, that LAMMPS forum message has been posted by me :) My collaborator @eokuznetsov is finishing his work on this HIP port. Currently, it covers all capabilities of the CUDA variant except for the neighbour list building on GPUs. Following your reply, we will wait until #1430 is merged and then plan to submit the pull request.

The performance of this HIP-backend is better than the performance of the OpenCL-backend: 
for ApoA1 we have got 10.7 ns/day vs 8.0 ns/day on Radeon VII with an 8 core Epyc CPU and ROCm ver.2.2.

**Comment by akohlmey** [comment]
> @akohlmey, actually, that LAMMPS forum message has been posted by me :) My collaborator @eokuznetsov is finishing his work on this HIP port. Currently, it covers all capabilities of the CUDA variant except for the neighbour list building on GPUs. Following your reply, we will wait until #1430 is merged and then plan to submit the pull request.

FYI, #1430 is merged.

**Comment by trinayan** [comment]
What is the current way to get the HIP version to build and run?.

**Comment by akohlmey** [comment]
This should be resolved with PR #2603
